### DISCUSSION TITLE
Generating Swagger for XML Requests?

### DISCUSSION THREAD

**Discussion by anzhi0708** [fastapi/fastapi#12476] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# I created 2 endpoints and the `/xml` endpoint accepts XML request. 
# I had to use `body: str = Body(..., media_type='application/xml')` as the 2nd param like this


from fastapi import FastAPI, Request, Response, Body
from fastapi.openapi.utils import get_openapi
from pydantic import BaseModel

app = FastAPI(title="Hello")


class _AccessIdentifier(BaseModel):
    SenderIdentifier: str
    ReceiverIdentifier: str

class _root(BaseModel):
    AccessIdentifier: _AccessIdentifier

class wraproot(BaseModel):
    root: _root


@app.post("/xml", response_class=Response)
async def xml(req: Request, body: str = Body(..., media_type='application/xml')):
    return Response(content="<result>OK</result>", media_type='application/xml')


@app.post("/")
async def main(request: Request, model: wraproot):
    return '{"result": "OK"}'


# After declaring the second parameter `body` as a `str`, 
# the server can receive and parse XML requests normally. 
# However, the swagger/openapi documentation generated at `localhost:8000/docs` cannot parse the schema of this `xml` endpoint. 
# If I change the type of the body parameter to `_root`:
```python
async def xml(req: Request, body: _root = Body(..., media_type='application/xml')): ...
```

Then /docs can generate the XML example and schema correctly, but at this point, the XML request will throw the following error: `{"detail":[{"type":"model_attributes_type","loc":["body"],"msg":"Input should be a valid dictionary or object to extract fields from","input":"<_root><AccessIdentifier><SenderIdentifier>ANJI</SenderIdentifier><ReceiverIdentifier>BROS</ReceiverIdentifier></AccessIdentifier></_root>"}]}`. 
This seems to create a paradox, where I can't have both benefits at the same time. Is there a solution to this?

### Description

I want the server to correctly receive XML requests while also generating the OpenAPI XML documentation properly. I need its documentation because I want to use tools like `schemathesis` for fuzz testing

### Operating System

macOS

### FastAPI Version

0.115.2

### Pydantic Version

2.9.2

### Python Version

3.11.5

**Comment by Kfir-G** [comment]
To solve this, use xmltodict to manually parse the XML body into a Python dict and convert it to your Pydantic model. This lets you handle XML requests while also maintaining OpenAPI docs.

```python
import xmltodict
from fastapi import FastAPI, Body, Response, HTTPException
from pydantic import BaseModel

app = FastAPI()

class _AccessIdentifier(BaseModel):
    SenderIdentifier: str
    ReceiverIdentifier: str

class _root(BaseModel):
    AccessIdentifier: _AccessIdentifier

class wraproot(BaseModel):
    root: _root

def parse_xml(body: str):
    try:
        return wraproot(**xmltodict.parse(body))
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Invalid XML: {e}")

@app.post("/xml", response_class=Response)
async def xml_endpoint(body: str = Body(..., media_type='application/xml')):
    parsed_body = parse_xml(body)
    return Response(content="<result>OK</result>", media_type="application/xml")
```

This keeps XML parsing and OpenAPI docs intact.
